symbolic-minidump/src/processor.rs

```diff
@@ -322,6 +331,195 @@ fn greater_than<A>(lower_bound: A) -> (Bound<A>, Bound<A>) {
     (Bound::Excluded(lower_bound), Bound::Unbounded)
 }
 
+/// The error type for [`SymbolSupplierError`]
+#[derive(Copy, Clone, Debug)]
+enum SymbolSupplierErrorKind {
+    NotFound,
+    NoDebugInfo,
+    SymCache,
+}
+
+impl fmt::Display for SymbolSupplierErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::NotFound => write!(f, "No symbol file found"),
+            Self::NoDebugInfo => write!(f, "File does not contain debug information"),
+            Self::SymCache => write!(f, "SymCache errror"),
+        }
+    }
+}
+
+/// An error encountered by a [`FileSystemSupplier`].
+#[derive(Debug, Error)]
+#[error("{kind}: {id}")]
+pub struct SymbolSupplierError {
+    id: CodeModuleId,
+    kind: SymbolSupplierErrorKind,
+    #[source]
+    source: Option<Box<dyn Error + Send + Sync + 'static>>,
+}
+
+/// A trait for types that can produce symcaches for given module id's,
+/// for example by reading debug information from the file system.
+pub trait SymCacheCreator<'a> {
+    /// Create a symcache for the given module.
+    fn create_symcache(
+        &self,
+        search_id: CodeModuleId,
+    ) -> Result<SelfCell<ByteView<'a>, SymCache<'a>>, SymbolSupplierError>;
+}
+
+/// A trait for types that can create and cache symcaches for given module
+/// id's.
+pub trait SymbolSupplier<'a> {
+    /// Return a reference to a symcache for the given module.
+    fn locate_symbols<'b: 'a>(
+        &'b mut self,
+        search_id: CodeModuleId,
+    ) -> Result<&'b SymCache<'a>, SymbolSupplierError>;
+}
+
+/// A [`SymbolSupplier`] that uses an internal [`SymCacheCreator`] to
+/// create symcaches and caches them in a map.
+pub struct SymCacheSupplier<'a, S> {
+    inner: S,
+    symcaches: SymCaches<'a>,
+}
+
+impl<'a, S> SymCacheSupplier<'a, S> {
+    /// Creates a new `SymCacheSupplier` from a given [`SymCacheCreator`].
+    pub fn new(inner: S) -> Self {
+        Self {
+            inner,
+            symcaches: SymCaches::new(),
+        }
+    }
+}
+
+impl<'a, S: SymCacheCreator<'a>> SymbolSupplier<'a> for SymCacheSupplier<'a, S> {
+    fn locate_symbols<'b: 'a>(
+        &'b mut self,
+        search_id: CodeModuleId,
+    ) -> Result<&'b SymCache<'a>, SymbolSupplierError> {
+        if let std::collections::btree_map::Entry::Vacant(e) = self.symcaches.entry(search_id) {
+            let symcache = self.inner.create_symcache(search_id)?;
```

Comment:
Maybe it would be a good idea to use the `or_insert_with` API, and really save a `Result` in the cache.